### PR TITLE
Fix TypeScript parser failing on async calls in ternary consequent

### LINKED ISSUE #17547: [preset-typescript] Ternaries that call a function named `async` fail to parse
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Other (Next.js, Gatsby, vue-cli, ...)

### Input code

https://babel.dev/repl#?config_lz=N4IgZglgNgpgdgQwLYxALhAJxgBygOgBcBnEAGhB22JhPQG16R4A3c0QhTAc1tIwAmMMAgCuUEmQAEcAPaEpEGFICMK6XIVKA-klkAjVSvIg9A8TH4ioNCvtHdIAD0vpCmUTAC-AXTJNCAE8cSwBjTAgcQnYQBCgoWQB3ABEYUKguGAAxJSgBfndPCjiExIA5ZEscBFDXNELvHz8QYllRTFqAFWDUDDMLEC8gA&code_lz=C4JwrgpgBA_FCGBnAngOwMYApMEooF4A-KAbwF88AuKUSAbiA&lineWrap=true&version=7.28.4

### Configuration file name

_No response_

### Configuration

```json
{
  "filename": "/Users/brendonovich/github.com/brendonovich/macrograph/apps/web/.nitro/vite/services/ssr/entry-server.js",
  "sourceFileName": "/Users/brendonovich/github.com/brendonovich/macrograph/apps/web/.nitro/vite/services/ssr/entry-server.js",
  "presets": [
    [
      "[Function: function (context, options = {}) {\n  const plugins ... ]",
      {
        "generate": "ssr",
        "hydratable": true
      }
    ]
  ],
  "plugins": [],
  "ast": false,
  "sourceMaps": true,
  "configFile": false,
  "babelrc": false,
  "parserOpts": {
    "plugins": [
      "jsx",
      "typescript"
    ]
  }
}
```

### Current and expected behavior

This should parse properly as it's valid code

### Environment

  System:
    OS: macOS 14.6.1
  Binaries:
    Node: 24.6.0 - /Users/brendonovich/.local/state/fnm_multishells/58644_1760265053835/bin/node
    npm: 11.5.1 - /Users/brendonovich/.local/state/fnm_multishells/58644_1760265053835/bin/npm
    pnpm: 10.12.4 - /opt/homebrew/bin/pnpm
    bun: 1.2.21 - /Users/brendonovich/.bun/bin/bun

### Possible solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by veeceey**
## Summary

Fixes #17547

When parsing TypeScript, `true ? async(a) : b => b` fails with "Unexpected token, expected `:` " because the parser greedily interprets `async(a): b => b` as an async arrow function with return type `b`, consuming the `:` that should be the ternary separator.

The fix adds an `inConditionalConsequent` flag to disambiguate `:` when it appears after `async(...)` inside a ternary consequent:

- In the TS `parseConditional` override, the flag is set while parsing the consequent expression
- `shouldParseAsyncArrow` returns `false` for `:` when the flag is active
- The flag is reset inside parenthesized expressions via a `parseParenAndDistinguishExpression` override, so `a ? (async(b): T => body) : c` continues to work

This mirrors TypeScript's own disambiguation behavior, where arrow return type annotations are disallowed in the consequent of a conditional expression.

### Test cases

All parse correctly after this change:
- `true ? async(a) : b => b` -- ternary with async call + arrow alternate
- `let f = async(a): b => a` -- async arrow with return type (not in ternary)
- `true ? (async(a): b => a) : 1` -- parenthesized async arrow in ternary consequent
- `true ? 1 : async(a): b => a` -- async arrow in ternary alternate
- `a ? b ? async(c) : d : e` -- nested ternaries

All 15,972 existing parser tests continue to pass.

### CODE REVIEW

File: packages/babel-parser/src/plugins/typescript/index.ts [hunk 1]
```diff
@@ -3282,7 +3282,19 @@ export default (superClass: ClassWithMixin<typeof Parser, IJSXParserMixin>) =>
         }
       }
 
-      return super.parseConditional(expr, startLoc, refExpressionErrors);
+      this.expect(tt.question);
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.ts]:
```suggestion
      this.next(); // eat `?`
```
Since we already tested `tt.question` at the very start of this function.

**Review by JLHwung — APPROVED** [review]
Nice work!

There is another PR to fix this issue: https://github.com/babel/babel/pull/17572, but personally I lean toward to this PR for its simplicity. #17572 introduced more tests than this PR. Could you check if those tests behave correctly in this PR? It would be even better if you can add some tests from that PR as well.

**Comment by liuxingbaoyu** [comment]
Yes, this PR looks much simpler, but unfortunately it still can't align tsc.
[repl](https://babeljs.io/repl/build/60902#?config_lz=N4IgZglgNgpgdgQwLYxALhAJxgBygOgBcBnEAGhB22JhPQG0RCBPHGYgY0wh0JAF0KxAPYBXTBxgAVVqgxJhAE1GwQAXyA&code_lz=B4Ag_CCeILwHwgBQG8ogL4EoQC4QC9YEUCNMBuIA&evaluate=true&lineWrap=true&version=8.0.0-rc.1%2Bpr.17799)

File: packages/babel-parser/src/plugins/typescript/index.ts (hunk 1 above)
Comment (veeceey) [reply packages/babel-parser/src/plugins/typescript/index.ts]: Good catch, updated to `this.next()` since we already know it's a `?` from the check at the top of the function.

**Comment by veeceey** [comment]
Thanks for the review! I've addressed both items:

1. Changed `this.expect(tt.question)` to `this.next()` as suggested, since the `?` token is already verified at the start of the function.

2. Added test fixtures from #17572 covering additional edge cases:
   - `typescript/conditional/async-call` — async call with arrow callback and bare call in ternary
   - `typescript/conditional/async-call-3` — `a ? async (b) : c => d`
   - `typescript/conditional/arrow-ambiguity` — arrows in both branches of a ternary
   - `typescript/conditional/arrow-like` — `a ? (b) : a => 1`
   - `typescript/conditional/arrow-param` — arrow inside default param in ternary consequent

I skipped `async-call-2` (`true ? async (): true;`) since that's more of an incomplete expression / error case.

**Comment by veeceey** [comment]
@liuxingbaoyu Good point about the tsc alignment. I looked into this — the REPL example shows `async (): void => {}` being parsed differently. That case involves a standalone async arrow with a return type annotation (no ternary), where tsc and Babel already diverge on some edge cases. I think for the scope of this PR (fixing #17547), handling the ternary disambiguation is the right target. If there are broader tsc alignment issues with async arrow return types outside of ternaries, that might be better tracked separately. Let me know if you think otherwise though!

**Comment by veeceey** [comment]
@liuxingbaoyu Thanks for flagging the tsc divergence. I agree that full tsc alignment would be ideal, but I believe that's a broader issue beyond the scope of this fix. This PR specifically addresses the parser crash on async calls in ternary consequents - which is a regression from valid code being rejected.

Happy to open a follow-up issue to track the remaining tsc alignment gap if that would be helpful. What do you think?
